src/market/directory.rs

```diff
@@ -234,35 +264,62 @@ pub fn start_directory_server(directory: Arc<DirectoryServer>) {
         .write(true)
         .create(true)
         .truncate(true) // Override the address file
-        .open(address_file.to_str().unwrap())
-        .unwrap();
-    file.write_all(file_content.as_bytes()).unwrap();
-    file.flush().unwrap();
+        .open(
+            address_file
+                .to_str()
+                .ok_or(DirectoryServerError::Other("Invalid address file path"))?,
+        )?;
+    file.write_all(file_content.as_bytes())?;
+    file.flush()?;
+
+    Ok(())
 }
 
 // The stream should have read and write timeout set.
 // TODO: Use serde encoded data instead of string.
-fn handle_client(stream: &mut TcpStream, addresses: Arc<RwLock<HashSet<String>>>) {
-    let reader_stream = stream.try_clone().unwrap();
+fn handle_client(
+    stream: &mut TcpStream,
+    directory: &Arc<DirectoryServer>,
+) -> Result<(), DirectoryServerError> {
+    let reader_stream = stream.try_clone().map_err(DirectoryServerError::Io)?;
     let mut reader = BufReader::new(reader_stream);
     let mut request_line = String::new();
 
-    reader.read_line(&mut request_line).unwrap();
+    reader
+        .read_line(&mut request_line)
+        .map_err(DirectoryServerError::Io)?;
 
     if request_line.starts_with("POST") {
         let addr: String = request_line.replace("POST ", "").trim().to_string();
-        addresses.write().unwrap().insert(addr.clone());
+        directory
+            .addresses
+            .write()
+            .map_err(|_| DirectoryServerError::LockError)?
+            .insert(addr.clone());
         log::info!("Got new maker address: {}", addr);
+
+        let address_file = directory.data_dir.join("addresses.dat");
+        // Expensive i/o operations below; Implement drop trait for directory server objects
+        let mut file = OpenOptions::new()
+            .create(true)
+            .append(true)
+            .open(address_file)
+            .map_err(DirectoryServerError::Io)?;
+        writeln!(file, "{}", addr).map_err(DirectoryServerError::Io)?;
```

Comment:
We don't need to open the file again here. The addresses are already stored in memory in `self.addresses`. Just reading from that will be enough.